### DISCUSSION TITLE
Sta command documentation

### DISCUSSION THREAD

**Discussion by dwRchyngqxs** [YosysHQ/yosys#3596] [Q&A] [answered]
Hello, I am trying to extract static timing analysis results using the `sta` command but the documentation is missing some info.

What option does this command takes? Is it possible to get fewer information, only latest arrival time for instance? Is there an option to output it in a nice format like `stat` json output?

I assume the arrival times are computed using timing information provided in https://github.com/YosysHQ/yosys/blob/3ebc50dee4007f8cca4ffc0e850bc3e86f7641f4/techlibs/ecp5/cells_sim.v, am I right?

What is the unit for the arrival time?
(I know the timing analysis cannot be accurate before place and route)

**Comment by jix** [comment, marked as answer]
The `sta` command doesn't take any options at all, so it's not possible to adjust how the output is presented. It uses the timing information loaded into the current design, so to use `ecp5/cells_sim.v` you would precede `sta` with `read_verilog -specify +/ecp5/cells_sim.v` where `-specify` enables loading `specify ... endspecify` blocks.

I'm not sure about the units, as far as I can tell, it just uses the values in the specify blocks without ever performing any unit conversion.

(I haven't used the `sta` command, this is just what I got from a glance at the implementation, so I might have missed things)

**Reply by dwRchyngqxs** [reply]
Yosys errors when I try to load the timing specification using `read_verilog -specify +/ecp5/cells_sim.v`:
```
ERROR: Non-constant enable $memwr$\mem$/usr/local/bin/../share/yosys/ecp5/cells_sim.v:207$99394_EN in memory initialization $meminit$\mem$/usr/local/bin/../share/yosys/ecp5/cells_sim.v:207$99420.
```

**Reply by jix** [reply]
Which Yosys version are you using? This does work with the latest version from git, as well as with Yosys 0.21 (random version from a few months ago I happened to have at hand).

**Reply by dwRchyngqxs** [reply]
Yosys 0.23+38 (git sha1 f9db7c059, clang 10.0.0-4ubuntu1 -fPIC -Os)
~I will post the whole command sequence I used that generated this error because~ executing only this command does not produce any error, it does after synthesizing a design (EDIT: probably a bug in the design, see below).

**Reply by dwRchyngqxs** [reply]
I ran it on a smaller example, it did not produce any error. But now `sta` doesn't produce any useful result either: [sta.log](https://github.com/YosysHQ/yosys/files/10269696/sta.log)
The script I ran is:
```
read_verilog top.v;
synth_ecp5 -abc9 -nowidelut;
read_verilog -specify +/ecp5/cells_sim.v;
tee -o sta.log sta
```

**Reply by jix** [reply]
Ah, I think it wants you to use `read_verilog -lib -specify +/ecp5/cells_sim.v;` instead (i.e. adding `-lib`), so that it doesn't try to analyze the timing of the simulation model, but knows that the modules in `cells_sim.v` are primitives for which the specify data should be used instead.

**Reply by dwRchyngqxs** [reply]
Adding `-lib` did the trick. Thank you for your quick answer.

**Reply by Ravenslofty** [reply]
> I'm not sure about the units, as far as I can tell, it just uses the values in the specify blocks without ever performing any unit conversion.

By convention, specify units are in picoseconds.
